### PR TITLE
Provide A Retry Policy For Ribbon Clients

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Helps address #1290.  Related to spring-cloud/spring-cloud-commons#137.

### CODE REVIEW

**Comment by ryanjbaxter** [comment]
The circleci build is failing because the spring-cloud-commons changes are not in master yet, not sure if there is any way to fix that...

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    private SpringClientFactory clientFactory;
+
+    public RibbonLoadBalancedRetryPolicyFactory(SpringClientFactory clientFactory) {
+        this.clientFactory = clientFactory;
+    }
+    @Override
+    public LoadBalancedRetryPolicy create(final String serviceId, final LoadBalancerClient loadBalancerClient) {
+        final RibbonLoadBalancerContext lbContext = this.clientFactory
+                .getLoadBalancerContext(serviceId);
+        return new LoadBalancedRetryPolicy() {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java]: I'd put this in a separate class at least.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                return nextServerCount <= lbContext.getRetryHandler().getMaxRetriesOnNextServer() && canRetry(context);
+            }
+
+            @Override
+            public void close(LoadBalancedRetryContext context) {
+
+            }
+
+            @Override
+            public void registerThrowable(LoadBalancedRetryContext context, Throwable throwable) {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java]: Do nothing with `throwable`?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancerClient.java [hunk 3]
```diff
@@ -29,15 +27,17 @@
...
+import java.util.Collections;
+import java.util.Map;
 
 /**
  * @author Spencer Gibb
  * @author Dave Syer
+ * @author Ryan Baxter
  */
-public class RibbonLoadBalancerClient implements LoadBalancerClient {
+public class RibbonLoadBalancerClient implements LoadBalancerClient{
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancerClient.java]: format lost space before brace.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java (hunk 2 above)
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java]: Like what? I think it is being passed just for informational purposes

File: spring-cloud-netflix-core/pom.xml [hunk 4]
```diff
@@ -45,6 +45,14 @@
 			<optional>true</optional>
 		</dependency>
 		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-aop</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.retry</groupId>
+			<artifactId>spring-retry</artifactId>
+		</dependency>
```
Comment (spencergibb) [inline spring-cloud-netflix-core/pom.xml]: similar to commons, should these be optional?

Comment (spencergibb) [inline spring-cloud-netflix-core/pom.xml]: Seems like it should be much simpler than commons.

**Comment by ryanjbaxter** [comment]
@spencergibb forgot to push my changes, take a look again, sorry about that

**Pull request merged**